### ISSUE TITLE
Onset tweaking

### ISSUE THREAD

**Issue by kylemcdonald** [librosa/librosa#305]
I'd like to use librosa onsets for separating transients for sampling. I noticed the peaks are always a little later than where I'd usually chop up a sound, so I wrote a little code to walk them back to the most recent local minima:

``` python
def tweak_onsets(onset_frames, o_env):
    tweaked = []
    for onset_frame in onset_frames:
        while True:
            if onset_frame == 0:
                break
            cur = o_env[onset_frame]
            prev = o_env[onset_frame - 1]
            if prev > cur:
                break
            onset_frame = onset_frame - 1
        tweaked.append(onset_frame)
    return np.array(tweaked)
```

Some onsets before:

![before](https://cloud.githubusercontent.com/assets/157106/13067714/6cbb68c2-d441-11e5-9a05-fe0e2f3e3ecb.png)

And after:

![after](https://cloud.githubusercontent.com/assets/157106/13067718/746bb2de-d441-11e5-9120-6c887c2cd3ed.png)

It's not exactly right, but for my application it works pretty well. I'm not sure that this is something librosa is designed for, and therefore not a bug per se, but I can imagine someone else wanting to use librosa to chop up sounds into transient samples, so I thought I'd leave this note here.

**Comment by bmcfee** [comment]
Interesting.. I'd need to see some quantitative results before adding something like this to the library, but it's worth considering.

A couple of alternate ideas:

1. Would it suffice to just use smaller hop sizes, and maybe `lag>1` (ala superflux)?
2. What about doing [time reassignment](https://www.researchgate.net/profile/Malcolm_Macleod/publication/2895902_Analysis_Of_Reassigned_Spectrograms_For_Musical_Transcription/links/00b4951856c4822ae9000000.pdf) on the detected frames?

For point 2, we already have an implementation of frequency reassignment, but it's buried inside `ifgram`.  It might be worth refactoring a bit to expose it (along with time reassignment), if there are more general use cases out there.

**Comment by bmcfee** [comment]
FWIW, I took a stab at implementing time reassignment [here](https://gist.github.com/bmcfee/a16c1a15b4f58baa3caf).

**Comment by bmcfee** [comment]
Pinging @kylemcdonald -- did you ever get a chance to try this out?

**Comment by Cortexelus** [comment]
I read the [time reassignment paper](https://www.researchgate.net/profile/Malcolm_Macleod/publication/2895902_Analysis_Of_Reassigned_Spectrograms_For_Musical_Transcription/links/00b4951856c4822ae9000000.pdf), and your gist, but I'm still confused by the reasoning/derivation behind this: 

`correction = np.mean((Dh / D).real, axis=0)`

![image](https://cloud.githubusercontent.com/assets/2453861/18602920/30d7a436-7c3b-11e6-8d79-dc977f89d64e.png)

The paper doesn't derive it, so now I'm reading the other papers it cites.

**Comment by bmcfee** [comment]
Okay, I'm convinced that we should implement the backtracking-to-minimum as a flag to the onset detector.

Time reassignment would also be nice to have, but I think it's solving a different problem than the one @kylemcdonald originally pointed out.

**Comment by Cortexelus** [comment]
I think what we want is backtracking to minimum loudness.

From the thesis of Tristan Jehan (EchoNest)  (posted in the related [amen issue](https://github.com/algorithmic-music-exploration/amen/issues/48))

![image](https://cloud.githubusercontent.com/assets/2453861/18606055/cd0c68ec-7c71-11e6-818b-8f371e7e1ae2.png)
![image](https://cloud.githubusercontent.com/assets/2453861/18606059/ea87841a-7c71-11e6-9cc1-70159478fdfd.png)

Since it returns sample indices, not frame indices, it should be a new function instead of a flag to the onset detector

**Comment by bmcfee** [comment]
> Since it returns sample indices, not frame indices, it should be a new function instead of a flag to the onset detector

Not sure I agree.  "Loudness" is not a sample-wise property, so it will have to be estimated over frames.  You can later back out the frame-center indices to the closest zero-crossing sample, but that's independent of the loudness backtracking.

**Comment by Cortexelus** [comment]
Okay good point. A separate find-nearest-zero-crossing function would be useful for other things too. It might have two parameters
1. look forward vs look backward vs find nearest 
2. pos-to-neg vs neg-to-pos vs whichever is closest

**Comment by bmcfee** [comment]
![We've already got one!](http://media.tumblr.com/tumblr_lp2pfjY08o1qjcv3k.gif)

[`effects.remix`](http://librosa.github.io/librosa/_modules/librosa/effects.html#remix) implements this via [`util.match_events`](http://librosa.github.io/librosa/generated/librosa.util.match_events.html).

It wouldn't be hard to separate this out and make it more efficient, using @tkell's implementation in [amen](https://github.com/algorithmic-music-exploration/amen/blob/master/amen/timing.py#L26-L90).  That one also handles stereo gracefully, but requires a bit more API/object overhead because it can have different zc offsets for each channel.

**Comment by kylemcdonald** [comment]
The time reassignment stuff is interesting, but yes I think it's solving a different problem.

Fundamentally what I was looking for (and what I'm guessing other remix-oriented people are looking for) is a feature akin to Ableton's automatic slicing functionality. From experience, I feel like Ableton is sample-based rather than frame-based in its final decision. `util.match_events` would be useful, but I think looking backward only (not nearest) is essential when talking about onsets.

Regarding zero-crossings, I can think of at least two counter examples: audio with a significant DC offset (rare) and mid/high frequency sounds where they oscillate/zc faster than their onset time (more common).

In short, I feel like an Ableton-quality slicer is not trivial, but you could get close looking for the minimum in a higher resolution "moving loudness" sample-accurate estimate.

Sorry to revive this thread I just saw it in my GitHub mentions and had to say something :)

**Comment by bmcfee** [comment]
Thanks @kylemcdonald for responding.

I'm thinking about the API here, and it's not completely obvious how best to implement this.

One option is to add an `energy` argument to `onset_detect`.  If `energy` is given, then backtracking to local min is enabled.  Otherwise, we leave it as is.  Note: we can't directly compute energy from within `onset_detect` because it's input may just be the onset envelope, and not the input signal.  

Backtracking on the onset envelope itself is probably not correct, but maybe better than nothing?  Should we make that an option as well?

Alternately, we could just make backtracking into a separate function that takes in onset times and an energy signal, and does the backtracking.

[Tagging @craffel for thoughts]

**Comment by craffel** [comment]
Hmm, backtracking is not something I have thought about or worked on.  I can understand why it's useful though.  Seems like it makes most sense to factor it into a separate function, in case someone wants to use it with their hand-rolled onset detector, unless that results in redundant computation.

**Comment by bmcfee** [comment]
Compromise: factor into it's own function, and optionally call it from `onset_detect`?

- `onset_backtrack(events, energy)` returns `events` (frame indices) rolled back to the most recent local min of `energy`
- `onset_detect(..., backtrack=True, energy=None)` calls above with `energy=onset_envelope`
- `onset_detect(..., backtrack=True, energy=e)` calls above with `energy=e` as a pre-computed energy curve `e` (eg, output of `feature.rmse`)

Sound reasonable?

**Comment by Cortexelus** [comment]
There should also be an example, if not its own function, that implements a
full  segmenter (Tristan's method)

 1) finding neg-to-pos* zero crossings
2) using util.match_events to align with back-tracked onsets.
3) works with stereo or mono

* sign-change needs to be consistent

**Comment by bmcfee** [comment]
> There should also be an example, if not its own function, that implements a
> full  segmenter (Tristan's method)

That's out of scope for this issue.

**Comment by Cortexelus** [comment]
I implemented a segmenter initially with the goal of getting it as close to echonest segments as possible. My segmenter began by following Tristan's description, then further optimizing it. To compare segmenters, I use librosa.util.match_events to pair up start_times. This throws out outliers. So I only compare events they have in common, and use that to measure error (absolute value of time difference). Here is an example of one rap song:

```python
# Optimized for accuracy, memory, and time
sr=22050, n_fft=512, win_length=256, hop_length=64, n_mels=25, mono=True
```

### CJ vs Echonest 
```python
Echonest Segments: 836
CJ Segments: 843
Extra Segments: 0%

# CJ Segments (no Loudness Backtracking)
Mean Segment Error in Frames:  18.7492368995
Mean Segment Error in Seconds:  0.0544188114619

# CJ Segments (with Loudness Backtracking)
Mean Segment Error in Frames:  7.58934889706
Mean Segment Error in Seconds:  0.0220277416602
```

### Librosa vs Echonest:
```python
Librosa Onsets: 290
Echonest Segments: 836
Extra Segments: 65%

# Librosa Default Onset Detect
Mean Segment Error in Frames:  18.0007270132
Mean Segment Error in Seconds:  0.0522462953965

# Librosa Default Onsets with Loudness Backtracking
Mean Segment Error in Frames:  10.8117472147
Mean Segment Error in Seconds:  0.031380606923
``` 

### Comparison: Venetian Snares - Szerencsétlen [[youtube](https://www.youtube.com/watch?v=vg52OSGmg7g)]
![image](https://cloud.githubusercontent.com/assets/2453861/20848145/ccd25620-b89e-11e6-9b86-1bb2c553a029.png)
Click below for full graphs:
![image](https://cloud.githubusercontent.com/assets/2453861/20847996/459d3fda-b89e-11e6-8349-c5b8cd677b06.png)
![image](https://cloud.githubusercontent.com/assets/2453861/20848022/5800a112-b89e-11e6-9b9c-e5b2cf53741c.png)

### Conclusion
I looked at results from only a handful of songs. 
- Librosa's onset detect finds too few onsets. Its params can be optimized, but I didn't try.
- Loudness backtracking significantly improves accuracy
- CJ segmenter generally has a mean error of 22ms-25ms compared to EchoNest.
- I didn't try shifting to local zero crossings, because: 
     - Windowing is an elegant solution for clicks
     - The stereo sample (1, -1) is a zero crossing in mono, but not a good segment boundary. 
     - Finding zero crossings in common for both channels are too rare, and will mess up the accuracy.
     - Using different zc per channel is an interesting thought, but adds complexity I'm not interested in supporting.
     - Show me otherwise though.
- Goal isn't to copy EchoNest exactly, but to make a robust segmenter that behaves how you'd expect, primarily for segment-based resynthesis, secondarily for analysis.

**Comment by bmcfee** [comment]
Closed by merging #473
